### PR TITLE
ASoC: SOF: relax PCM period size constraints

### PR DESCRIPTION
**Pull request by kv2019i**
Current SOF implementation limits period size to multiples
of period_min. Period_min is defined in topology, but is in practise
set to align with the SOF DSP timer tick (typically 1ms).

While this approach helps user-space to avoid period sizes, which are
not aligned to the DSP timer tick, it causes problems to applications
which want to align data processing size to that of ALSA period
size. One example is JACK audio server, which limits period sizes
to power of two values.

Other ALSA drivers where audio data tranfer is driven by a timer
tick, like USB, do not enforce the timer tick to all period sizes.

To align SOF to follow the same behaviour, drop the additional
alignment constraint on period_size.

BugLink: https://github.com/thesofproject/linux/issues/2391
Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

**Comment by kv2019i** [comment]
Tested with JACK and with this patch it works out of the box without the tedious exercise to find a 2^n buffer size.

CI results are also green. This patch does result in different period sizes to be chosen in a large number of test cases and real-life use-cases, so maybe extra level care is needed to check we have no regressions, but so far seems pretty good.

File: sound/soc/sof/pcm.c [hunk 1]
```diff
@@ -482,9 +482,6 @@ static int sof_pcm_open(struct snd_soc_component *component,
 	snd_pcm_hw_constraint_step(substream->runtime, 0,
 				   SNDRV_PCM_HW_PARAM_BUFFER_BYTES,
 				   le32_to_cpu(caps->period_size_min));
-	snd_pcm_hw_constraint_step(substream->runtime, 0,
-				   SNDRV_PCM_HW_PARAM_PERIOD_BYTES,
-				   le32_to_cpu(caps->period_size_min));
```
Comment (plbossart) [inline sound/soc/sof/pcm.c:487]: with this you remove ALL constraints on the period size.
I would assert the period size needs to be at least caps->period_size_min.

You also need to add in your commit message that choosing such periods will results in non-periodic events reported, and while it may be nice for some applications it may be problematic if the application checks when the period is reported.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/pcm.c:487]: @plbossart wrote
> I would assert the period size needs to be at least caps->period_size_min

The minimum is set a few lines down in pcm_open:
runtime->hw.period_bytes_min = le32_to_cpu(caps->period_size_min);

But, but, I now wonder should we modify the buffer size constraint as well. This could result in weird period/buffer size combos, if we restrict only one.

> You also need to add in your commit message that choosing such periods will results in non-periodic events reported

I tried to capture this with "While this approach helps user-space to avoid..." sentence. OTOH, if any application is surprised by this, then it will be surprised by any USB audio device as well.

Comment (plbossart) [reply sound/soc/sof/pcm.c:487]: yes, I was wondering how you can have a buffer that is required to be a multiple of 1ms while allowing for periods that are not?

Comment (kv2019i) [reply sound/soc/sof/pcm.c:487]: @plbossart Upon some further investigation, I see now reason to constraint the buffer size either, other than min/max. Patch and commit message updated.

**Comment by kv2019i** [comment]
@dbaluta @paulstelian97 Can you check as well as these constraints are in common SOF code and apply to all platforms.

**Comment by paulstelian97** [comment]
> @dbaluta @paulstelian97 Can you check as well as these constraints are in common SOF code and apply to all platforms.

I think our topologies still limit period size anyway (to a fixed value)?

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/pcm.c:487]: @kv2019i don't we want the buffer size to be a multiple of the period size? I can't recall what this constraint is but that seems rather useful to enforce, no?

Comment (kv2019i) [reply sound/soc/sof/pcm.c:487]: @plbossart Yes, been looking at that as well. That seems like a very reasonable constraint (i.e. any app that uses period-based wake-ups, probably wants an even multiple), but its use in drivers seem somewhat adhoc. I'm not really convinced drivers should add this constraint unless it is something the hw enforces.

For the HDA implementation in SOF, we even have code in hda_dsp_stream_setup_bdl() to explicitly handle the case where buffersize is not a multiple of period size. OTOH, legacy HDA driver does set the integer constraint and there is some code in legacy driver that at least seems suspicious (e.g. snd_hdac_stream_setup_periods(), SOF is not using this). But as a good example, the HDA legacy driver line that added the constraint is from a commit done in 2005, with no rationale why integer periods is used -- just a limitation of current code, or something in the hw.

So adding this at SOF common level does not seem right. There are many widely used drivers that do not set such constraint and applications work fine with them.

Let me stress test this a bit more.

Comment (plbossart) [reply sound/soc/sof/pcm.c:487]: @kv2019i maybe it's true for HDAudio but I am not sure this would work for SOF in general, e.g. for BYT/BDW legacy platforms where the DMA is handled by the firmware. To the best of my knowledge those DMAs only handle linear memory fragments, and if you have a period straddling across the end and the start of the ring buffer we may have a problem.

**Comment by kv2019i** [comment]
@plbossart How about this? I now moved the Intel HW specific constraints to Intel platform specific code, and only kept the constraints coming from topology, in the common SOF PCM code. I aligned the constraints with legacy driver and kept the integer period count also HDA platforms. Legacy driver seems to have it as well.

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/pcm.c:487]: To address this, this constraint is added back in 2nd Intel specific patch.

File: sound/soc/sof/intel/intel-ipc.c [hunk 2]
```diff
@@ -73,6 +73,13 @@ int intel_pcm_open(struct snd_sof_dev *sdev,
 	/* binding pcm substream to hda stream */
 	substream->runtime->private_data = stream;
 
+	/* minimum as per HDA spec */
+	snd_pcm_hw_constraint_step(substream->runtime, 0, SNDRV_PCM_HW_PARAM_PERIOD_BYTES, 4);
+
+	/* avoid circular buffer wrap in middle of period */
+	snd_pcm_hw_constraint_integer(substream->runtime,
+				      SNDRV_PCM_HW_PARAM_PERIODS);
+
```
Comment (plbossart) [inline sound/soc/sof/intel/intel-ipc.c:82]: why do we have the same stuff in two places? This doesn't seem to be related to HDAudio?

Also we're discussing in https://github.com/thesofproject/sof/issues/3503 and https://github.com/thesofproject/sof/issues/3492 that Broadwell seems to have more restrictions on the DMA than BYT/CHT, so we might need to make this really platform-specific.

**Review by plbossart — COMMENTED** [review]
Sounds better @kv2019i but I am confused on the intel_pcm_open(), see below.

We might also require @dbaluta to add constraints for i.MX if needed?

**Comment by kv2019i** [comment]
> We might also require @dbaluta to add constraints for i.MX if needed?

Ack, I pinged them already earlier and response was that topology constraints are sufficient. But correct, in general, any HW specific constraints would have to be added for each platform.

File: sound/soc/sof/intel/intel-ipc.c (hunk 2 above)
Comment (kv2019i) [reply sound/soc/sof/intel/intel-ipc.c:82]: BYT/CHT/etal use intel_pcm_open() while APL and newer call hda_dsp_pcm_open(). As it indeed sounds like BYT/CHT might have even more constraints, it seemed best to keep these separate and have a clear place where to add the constraints.

The integer constraint is not HDA specific, but it's something we have applied for Intel hw in the past and we have at least some Intel platforms where this is likely mandatory. I still don't see a need to make this constraint in common SOF code as that will limit all platforms.

Comment (plbossart) [reply sound/soc/sof/intel/intel-ipc.c:82]: you may want to remove the reference to HDAudio in the non-HDaudio case then

Comment (kv2019i) [reply sound/soc/sof/intel/intel-ipc.c:82]: ack @plbossart , fixed now

**Review by plbossart — APPROVED** [review]
thanks @kv2019i

**Comment by kv2019i** [comment]
@keyonjie @bardliao @lyakh @juimonen @ranj063 can you take a look?

**Comment by ranj063** [comment]
> @keyonjie @bardliao @lyakh @juimonen @ranj063 can you take a look?

@kv2019i the changes look good but shouldnt we squash the 2 patches?

**Comment by kv2019i** [comment]
@ranj063 wrote:
> @kv2019i the changes look good but shouldnt we squash the 2 patches?

They are somewhat separate now and easier to explain separately. We are not adding the same constraints back we remove from common code, back to Intel specific code.

**Review by ranj063 — APPROVED** [review]

**Comment by ranj063** [comment]
> @ranj063 wrote:
> 
> > @kv2019i the changes look good but shouldnt we squash the 2 patches?
> 
> They are somewhat separate now and easier to explain separately. We are not adding the same constraints back we remove from common code, back to Intel specific code.

I still feel like they shouldbe squashed but given that the even the first commit doesnt cause any regressions, it might be easier to keep them separate to make it easier to follow

**Comment by kv2019i** [comment]
Thanks for reviews, I'll push this in now to expand the testing coverage.
